### ISSUE TITLE
System.FormatException: Input string was not in a correct format

### ISSUE THREAD

**Issue by DaveSkender** [facioquo/stock-indicators-python#382]
### The problem

We're still seeing users struggle with `locale` related conversion problems for quotes.  It's certainly due to a misalignment of environment settings; however, I think there's more we can do to better understand it and possibly find more solutions to either mitigate the problem or to help users know exactly what they need to change when it does occur, maybe a _self-test_ that they can run locally.

_See original post by @ArtemHachaturov in https://github.com/DaveSkender/Stock.Indicators/discussions/1194_ for more information and source quotes.

**Comment by DaveSkender** [comment]
There's also this `<NeutralLanguage>en-US</NeutralLanguage>` tag [hard-coded into the .NET library](https://github.com/DaveSkender/Stock.Indicators/blob/b826b3bc420299dc0797235e1c96cb3a89d11bb5/src/Indicators.csproj#L18), but am not sure if it's a problem.

**Comment by ArtemHachaturov** [comment]
Yes, I agree with you, of course, but I just want to solve my problem

**Comment by LeeDongGeon1996** [comment]
After quick search, I found that we might convert Decimal into str incorrectly. Looks like just using `str()` doesn't affected by  `locale` setting. 
See here: https://stackoverflow.com/questions/31910741/how-can-i-locale-format-a-python-decimal-and-preserve-its-precision

**Comment by LeeDongGeon1996** [comment]
Hi, @ArtemHachaturov . Thanks for your detail report. As I mentioned above, it looks like the `locale` setting does not apply in converting into `str`
Can you give me the results from the code below?
```python
from decimal import Decimal
value = Decimal(12.345)
print(value) # case 1
print('{:n}'.format(value)) # case 2
print(format(value, 'n')) # case 3

from stock_indicators._cslib import CsDecimal
print(CsDecimal.Parse(format(value, 'n'))) # case 4
```

**Comment by ArtemHachaturov** [comment]
There is a small advance, the values fall into quotes 


import pandas as pd
from stock_indicators import indicators, Quote
from babel.numbers import format_decimal

df = pd.read_csv('quotes.csv', parse_dates=['Time'])

def format_with_babel(number, locale='ru_RU'):
    return format_decimal(number, locale=locale)

df['Open'] = df['Open'].apply(format_with_babel)
df['High'] = df['High'].apply(format_with_babel)
df['Low'] = df['Low'].apply(format_with_babel)
df['Close'] = df['Close'].apply(format_with_babel)
df['Volume'] = df['Volume'].apply(lambda x: format_with_babel(x, locale='ru_RU'))

quotes = [
    Quote(date, open, high, low, close, volume)
    for date, open, high, low, close, volume
    in zip(df['Time'], df['Open'], df['High'], df['Low'], df['Close'], df['Volume'])
]
print(quotes)

#[<stock_indicators.indicators.common.quote.Quote object at 0x0000023D856F24C0>, <stock_indicators.indicators.common.quote.Quote object at 0x0000023D88E87080>....


print("Time       | Open | High | Low  | Close | Volume")
for q in quotes:
    print(f"{q.date:%Y-%m-%d} | {q.open} | {q.high} | {q.low} | {q.close} | {q.volume})")

#But at the end there is an error

Time       | Open | High | Low  | Close | Volume
Traceback (most recent call last):
  File "e:\Binance Bot\python\12.py", line 34, in <module>
    print(f"{q.date:%Y-%m-%d} | {q.open} | {q.high} | {q.low} | {q.close} | {q.volume})")
                                 ^^^^^^
  File "C:\Users\russt\AppData\Local\Programs\Python\Python311\Lib\site-packages\stock_indicators\indicators\common\quote.py", line 22, in _get_open
    return to_pydecimal(quote.Open)
           ^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\russt\AppData\Local\Programs\Python\Python311\Lib\site-packages\stock_indicators\_cstypes\decimal.py", line 36, in to_pydecimal
    return PyDecimal(str(cs_decimal))
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
decimal.InvalidOperation: [<class 'decimal.ConversionSyntax'>]

**Comment by LeeDongGeon1996** [comment]
@ArtemHachaturov 
Ohhhh! Thank you. That was very helpful to figure out what's the problem.
Could you give me more information? It looks like your python runtime is using the locale, which uses period(`.`) as a decimal point. But, your DotNet runtime is using comma(`,`) as a decimal point.
If you can give me the results after running the below, It'll be very appreciated.

On Python:
```python
import locale
print (locale.getlocale(locale.LC_ALL))
print (locale.localeconv())
```

On your terminal:
```bash
locale
```

**Comment by ArtemHachaturov** [comment]
Dear friends, I am glad to inform you that everything worked out for me and now everything works like a clock.

I had only one locale, this was RU, I understood that I still needed to install the language pack us, but for certain reasons I was unable to do this until I completely demolished the system and installed another Windows, where I also tried to download it manually the language pack was installed but was not applied until I used the following commands
 
 - cmd
 control intl.cpl,, /f:"%SystemRoot%\System32\LangRes.dll",/s:"0x00000409"
 
 -PowerShell
 Set-WinSystemLocale en-US
 Set-Culture en-US

and restarting the computer

import locale

current_locale = locale.getdefaultlocale()[0]

print(current_locale) #en_US

I realized that everything should work!

Everything looks amazing

df = pd.read_csv('quotes.csv', parse_dates=['Time'])
quotes = [
    Quote(date, open, high, low, close, volume)
    for date, open, high, low, close, volume
    in zip(df['Time'], df['Open'], df['High'], df['Low'], df['Close'], df['Volume'])
]

print("Time       | Open | High | Low  | Close | Volume")
for q in quotes:
    print(f"{q.date:%Y-%m-%d} | {q.open} | {q.high} | {q.low} | {q.close} | {q.volume})")
    
2018-07-27 | 275.57 | 275.68 | 272.34 | 273.35 | 79050080)
2018-07-30 | 273.44 | 273.61 | 271.35 | 271.92 | 65624404)
2018-07-31 | 272.76 | 273.93 | 272.34 | 273.26 | 70594928)
2018-08-01 | 273.49 | 274.04 | 272.1 | 272.81 | 55443260)
2018-08-02 | 271.38 | 274.48 | 271.15 | 274.29 | 65298924)
2018-08-03 | 274.43 | 275.52 | 274.23 | 275.47 | 55527740)
2018-08-06 | 275.51 | 276.82 | 275.08 | 276.48 | 40564136)
2018-08-07 | 277.21 | 277.81 | 277.06 | 277.39 | 44471960)

Thank you very much to you and your team, I wish you great success and prosperity

**Comment by DaveSkender** [comment]
I suspect this essentially reset your PC environment locale to `en_US`, which would certainly solve your blocking issue; however, we’re still going to try and get it working better on other locales.

**Comment by SnowyOw1** [comment]
I am able to reproduce the symptom. Resetting system locale is clearly not an option in my case.

Indeed [the problem is at](https://github.com/facioquo/stock-indicators-python/blob/main/stock_indicators/_cstypes/decimal.py#L36)
`CsDecimal.Parse(str(decimal))`

> ```python
> import locale
> print(locale.localeconv())
> print(locale.getlocale(locale.LC_ALL))
> ```

```
{'int_curr_symbol':` '', 'currency_symbol': '', 'mon_decimal_point': '', 'mon_thousands_sep': '', 'mon_grouping': [], 'positive_sign': '', 'negative_sign': '', 'int_frac_digits': 127, 'frac_digits': 127, 'p_cs_precedes': 127, 'p_sep_by_space': 127, 'n_cs_precedes': 127, 'n_sep_by_space': 127, 'p_sign_posn': 127, 'n_sign_posn': 127, 'decimal_point': '.', 'thousands_sep': '', 'grouping': []}
```

>   File "<input>", line 1, in <module>
>   File "C:\Program Files\Python311\Lib\locale.py", line 608, in getlocale
>     raise TypeError('category LC_ALL is not supported')

**Comment by LeeDongGeon1996** [comment]
@SnowyOw1 Thank you for reporting. We are fixing it now. Could you try with this patched version? (it's not merged yet but about to be.)

```bash
pip uninstall stock-indicators
pip install git+https://github.com/LeeDongGeon1996/Stock.Indicators.Python.git@fix/apply-locale
```

**Comment by LeeDongGeon1996** [comment]
Patched version released on test PyPI. Please have a test and any feedback appreciated.
```bash
pip install -i https://test.pypi.org/simple/ stock-indicators==1.3.1.dev9
```

**Comment by SnowyOw1** [comment]
> @SnowyOw1 Thank you for reporting. We are fixing it now [here](https://github.com/facioquo/stock-indicators-python/pull/386). Could you try with this patched version? (it's still in reviewing yet but about to be merged.)
> 
> +and If you let me know your `locale` setting on your machine, it'll be very appreciated.

I hereby confirrm the fix is working. I didn't apply the full patch but monkey-patched sources.

My locale is

```
PS C:\Users\fedrushkovv> Get-WinSystemLocale

LCID             Name             DisplayName
----             ----             -----------
1049             ru-RU            Russian (Russia)


PS C:\Users\fedrushkovv> (Get-WinSystemLocale).NumberFormat


CurrencyDecimalDigits    : 2
CurrencyDecimalSeparator : ,
IsReadOnly               : False
CurrencyGroupSizes       : {3}
NumberGroupSizes         : {3}
PercentGroupSizes        : {3}
CurrencyGroupSeparator   :  
CurrencySymbol           : ₽
NaNSymbol                : не число
CurrencyNegativePattern  : 8
NumberNegativePattern    : 1
PercentPositivePattern   : 1
PercentNegativePattern   : 1
NegativeInfinitySymbol   : -∞
NegativeSign             : -
NumberDecimalDigits      : 2
NumberDecimalSeparator   : ,
NumberGroupSeparator     :  
CurrencyPositivePattern  : 3
PositiveInfinitySymbol   : ∞
PositiveSign             : +
PercentDecimalDigits     : 2
PercentDecimalSeparator  : ,
PercentGroupSeparator    :  
PercentSymbol            : %
PerMilleSymbol           : ‰
NativeDigits             : {0, 1, 2, 3...}
DigitSubstitution        : None
```

**Comment by DaveSkender** [comment]
Thank you for confirming.  This patch has been published in [v1.3.1](https://pypi.org/project/stock-indicators/1.3.1/)
